File: app/src/main/java/social/entourage/android/map/filter/MapFilter.java
package social.entourage.android.map.filter;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

import social.entourage.android.api.model.TourType;
import social.entourage.android.api.model.map.Entourage;
import social.entourage.android.map.entourage.category.EntourageCategory;

/**
 * Created by mihaiionescu on 17/05/16.
 */
public class MapFilter implements Serializable {

    private static final long serialVersionUID = -2822136342813499636L;

    // ----------------------------------
    // Attributes
    // ----------------------------------

    public static final int DAYS_1 = 24; //hours
    public static final int DAYS_2 = 8*24; //hours
    public static final int DAYS_3 = 30*24; //hours

    public boolean tourTypeMedical = true;
    public boolean tourTypeSocial = true;
    public boolean tourTypeDistributive = true;

    public boolean entourageTypeDemand = true;
    public boolean entourageTypeContribution = true;

    List<String> entourageCategories = new ArrayList<>();

    public boolean showTours = true;

    public boolean onlyMyEntourages = false;
    public boolean onlyMyPartnerEntourages = false;

    public int timeframe = DAYS_2;

    // ----------------------------------
    // Lifecycle
    // ----------------------------------

    private static MapFilter ourInstance = new MapFilter();

    public static MapFilter getInstance() {
        return ourInstance;
    }

    protected MapFilter() {
    }

    // ----------------------------------
    // Methods
    // ----------------------------------

    public String getTypes() {
        StringBuilder entourageTypes = new StringBuilder("");

        if (tourTypeMedical) {
            entourageTypes.append(TourType.MEDICAL.getKey());
        }
        if (tourTypeSocial) {
            if (entourageTypes.length() > 0) entourageTypes.append(",");
            entourageTypes.append(TourType.BARE_HANDS.getKey());
        }
        if (tourTypeDistributive) {
            if (entourageTypes.length() > 0) entourageTypes.append(",");
            entourageTypes.append(TourType.ALIMENTARY.getKey());
        }
        for (String categoryKey: entourageCategories) {
            if (entourageTypes.length() > 0) entourageTypes.append(",");
            entourageTypes.append(categoryKey);
        }

        return entourageTypes.toString();
    }

    public boolean isCategoryChecked(EntourageCategory entourageCategory) {
        if (Entourage.TYPE_DEMAND.equals(entourageCategory.getEntourageType())) return entourageTypeDemand && entourageCategories.contains(entourageCategory.getKey());
        if (Entourage.TYPE_CONTRIBUTION.equals(entourageCategory.getEntourageType())) return entourageTypeContribution && entourageCategories.contains(entourageCategory.getKey());
        return false;
    }

    public void setCategoryChecked(String actionCategory, boolean checked) {
        if (checked) {
            if (!entourageCategories.contains(actionCategory)) {
                entourageCategories.add(actionCategory);
            }
        } else {
            entourageCategories.remove(actionCategory);
        }
    }

}

File: app/src/main/java/social/entourage/android/map/filter/MapFilter.kt
package social.entourage.android.map.filter

import social.entourage.android.api.model.map.BaseEntourage
import social.entourage.android.entourage.category.EntourageCategory
import social.entourage.android.entourage.category.EntourageCategoryManager
import java.io.Serializable
import java.util.*

/**
 * Created by mihaiionescu on 17/05/16.
 */
class MapFilter : MapFilterInterface, Serializable {
    var entourageTypeOuting = true
    var showPastEvents = false
    var entourageTypeDemand = true
    var entourageTypeContribution = true
    var timeframe = DAYS_3

    private var entourageCategories: MutableList<String> = ArrayList()

    // ----------------------------------
    // MapFilter implementation
    // ----------------------------------
    override fun getTypes(): String {
        val entourageTypes = StringBuilder()
        if (entourageTypeOuting) {
            if (entourageTypes.isNotEmpty()) entourageTypes.append(",")
            entourageTypes.append("ou")
        }
        for (categoryKey in entourageCategories) {
            if (entourageTypes.isNotEmpty()) entourageTypes.append(",")
            entourageTypes.append(categoryKey)
        }
        return entourageTypes.toString()
    }

    override fun getTimeFrame(): Int {
        return timeframe
    }

    override fun showPastEvents(): Boolean {
        return showPastEvents
    }

    override fun entourageCreated() {
        entourageTypeContribution = true
        entourageTypeDemand = true
    }

    override fun validateCategories() {
        if (entourageTypeDemand) validateCategoriesForGroup(BaseEntourage.GROUPTYPE_ACTION_DEMAND)
        if (entourageTypeContribution) validateCategoriesForGroup(BaseEntourage.GROUPTYPE_ACTION_CONTRIBUTION)
    }

    override fun isDefaultFilter(): Boolean{
        return when {
            !entourageTypeDemand  -> false
            !entourageTypeContribution -> false
            //normal filter
            !entourageTypeOuting -> false
            //TODO check tis one: //showPastEvents -> false
            timeframe != DAYS_3 -> false
            else -> true
        }
    }

    override fun setDefaultValues() {
        entourageTypeDemand = true
        entourageTypeContribution = true
        entourageTypeOuting=true
        showPastEvents=false
        timeframe = DAYS_3
    }

    // ----------------------------------
    // Methods
    // ----------------------------------
    fun isCategoryChecked(actionGroupType: EntourageCategory): Boolean {
        return when(actionGroupType.groupType) {
            BaseEntourage.GROUPTYPE_ACTION_DEMAND-> {
                entourageTypeDemand && entourageCategories.contains(actionGroupType.key)
            }
            BaseEntourage.GROUPTYPE_ACTION_CONTRIBUTION -> {
                entourageTypeContribution && entourageCategories.contains(actionGroupType.key)
            }
            else -> false
        }
    }

    fun setCategoryChecked(actionCategory: String, checked: Boolean) {
        if (checked) {
            if (!entourageCategories.contains(actionCategory)) {
                entourageCategories.add(actionCategory)
            }
        } else {
            entourageCategories.remove(actionCategory)
        }
    }

    // ----------------------------------
    // Serialization
    // ----------------------------------
    private fun validateCategoriesForGroup(actionGroup: String) {
        //get the list of categories for this type
        val categoryList = EntourageCategoryManager.getEntourageCategoriesForGroup(actionGroup)
        //search for the keys
        for(category in categoryList) {
            if (entourageCategories.contains(category.key)) {
                //allKeysFalse = false
                return
            }
        }
        //set all keys to true
        //search for the keys
        categoryList.forEach {it.key?.let {key -> entourageCategories.add(key)} }
    }

    companion object {
        private const val serialVersionUID = -2822136342813499636L

        // ----------------------------------
        // Attributes
        // ----------------------------------
        const val DAYS_1 = 24 //hours
        const val DAYS_2 = 8 * 24 //hours
        const val DAYS_3 = 30 * 24 //hours
    }

    init {
        validateCategories()
    }
}